Contribute page (#711)
adobe-private/react-spectrum-v3: packages/dev/docs/pages/contribute.mdx

---

# Contribute

> Thanks for choosing to contribute! We look forward to improving web applications together. Here you will find information on how propose bug fixes, suggest improvements and how to build react-spectrum yourself.

Reviewer: ```suggestion
Thanks for choosing to contribute! We look forward to improving web applications together. Here you will find information on how propose bug fixes, suggest improvements, and develop locally.
```

## Better together

> We believe in sharing our process, results and journey with you all. The core team and all external contributors follow the same process in order maintain a high quality codebase.

Reviewer: ```suggestion
We believe that the best way to build a better web is together as a community. The React Spectrum project aims to make it easier to build design systems and component libraries with high quality interactions and accessibility for all. The core team and all external contributors follow the same process in order maintain a high quality codebase.
```

## Code of conduct

> We adhere to the Adobe [code of conduct](https://github.com/adobe-private/react-spectrum-v3/blob/CODE_OF_CONDUCT.md) and by participating, you are expected to uphold this code. Please report unacceptable behavior to Grp-opensourceoffice@adobe.com.

Reviewer: Just noticed the link 404s

## Reporting issues
### Bugs

> We use [GitHub issues](https://github.com/adobe-private/react-spectrum-v3/issues) to track work and log bugs. Please check existing issues before filing anything new. We do our best to respond to issues within a few days. If you would like to contribute a fix, please let us know on the issue.

Reviewer: ```suggestion
We use [GitHub issues](https://github.com/adobe-private/react-spectrum-v3/issues) to track work and log bugs. Please check existing issues before filing anything new. We do our best to respond to issues within a few days. If you would like to contribute a fix, please let us know by leaving a comment on the issue.
```

> The best way to reduce back and forth on a bug is to add code for a storybook example exhibiting the issue along with steps to reproduce it. If you would like to work on a bugfix yourself, make sure an issue exists first.

Reviewer: I think we don't need it to be storybook specifically? Something like a code sandbox link would also work. Just need to be a minimal reproduction.

```suggestion
The best way to reduce back and forth on a bug is provide a small code example exhibiting the issue along with steps to reproduce it. If you would like to work on a bugfix yourself, make sure an issue exists first.
```

> 

Reviewer: Maybe add something asking that people follow the issue template like you have for PRs?

### Feature requests

> Our components can always be improved upon. If you have a feature request, you can use our Feature Request issue template in issues. For larger scopes of work, it is a good idea to put forth a Request For Comments (RFC) first to gather feedback from the team. A [template](https://github.com/adobe-private/react-spectrum-v3/blob/master/rfcs/template.md) is available to build an RFC. Make a PR to add your RFC to the [rfcs folder](https://github.com/adobe-private/react-spectrum-v3/tree/master/rfcs) to give the team and the community a chance to discuss the proposal.

Reviewer: ```suggestion
Our components can always be improved upon. If you have a feature request, you can use our Feature Request issue template. For larger scopes of work, it is a good idea to open a Request For Comments (RFC) first to gather feedback from the team. Please follow our RFC [template](https://github.com/adobe-private/react-spectrum-v3/blob/master/rfcs/template.md). Make a PR to add your RFC to the [rfcs folder](https://github.com/adobe-private/react-spectrum-v3/tree/master/rfcs) to give the team and the community a chance to discuss the proposal.
```

### Security issues

> Security issues shouldn't be reported on this issue tracker. Instead, [file an issue](https://helpx.adobe.com/security/alertus.html) to our security experts.

Reviewer: ```suggestion
Security issues shouldn't be reported on this issue tracker. Instead, please follow the directions [here](https://helpx.adobe.com/security/alertus.html) to contact our security team.
```

## Where to start
There are many places to dive into react-spectrum to help out. Before you take on a feature or issue, make sure you become familiar with [our architecture](architecture.html).

If you are looking for place to start, consider the following options:
- Look for issues tagged with help wanted and/or good first issue.
- Help triage existing issues by investigating problems and following up on missing information.
- Update missing or fix existing documentation
- Review and test open pull requests

## Developing
When you are ready to start developing you can clone the repo and start storybook.
Make sure you have the following requirements installed: [node](https://nodejs.org/) (v11.0.0+) and [yarn](https://yarnpkg.com/en/) (v1.22.0+)

Fork the repo first using [this guide](https://help.github.com/articles/fork-a-repo), then clone it locally.
```

> git clone https://github.com/YOUR-USERNAME/react-spectrum-v3

Reviewer: ```suggestion
git clone https://github.com/YOUR-USERNAME/react-spectrum
```

> cd react-spectrum-v3

Author: ```suggestion
cd react-spectrum
```

yarn install
```

You can then run the storybook and browse to [http://localhost:9003](http://localhost:9003) with:
```bash
yarn run
```

Or run the documentation and browse to [http://localhost:1234/](http://localhost:1234/) with:
```bash
yarn start:docs
```

### Tests

> We use [jest](https://jestjs.io/) for unit tests and [react-testing-library](https://testing-library.com/docs/react-testing-library/intro) for rendering and writing assertions. Please make sure you include tests with your pull requests. Our CI/CD will run the tests on PRs as well as lint code and TS. You can see on each PR whether you have passed all our checks. We split the tests into 2 groups.

Reviewer: ```suggestion
We use [jest](https://jestjs.io/) for unit tests and [react-testing-library](https://testing-library.com/docs/react-testing-library/intro) for rendering and writing assertions. Please make sure you include tests with your pull requests. Our CI will run the tests on PRs as well as the linter and type checker. You can see on each PR whether you have passed all our checks. We split the tests into 2 groups.
```

*Visual tests*

> - A Storybook story should be written for any visual breakage of a component.

Reviewer: ```suggestion
- A Storybook story should be written for each visual state that a component can be in (based on props).
```

*Unit tests*
- (Props) Anything that should be changed by a prop should be tested via react-testing-library.
- (Events) Anything that should trigger an event should be tested via react-testing-library.

You can run the tests with:

```bash
yarn jest
```

You can also get a code coverage report by running:

```bash
yarn jest --coverage
```

### Linting
The code is linted with [eslint](https://eslint.org/). The linter runs whenever you run the tests, but you can also run it with
```bash
yarn lint
```

### TypeScript
The code for React Spectrum is written in [TypeScript](https://www.typescriptlang.org/). The type checker will usually run in your editor, but also runs when you run
```bash
yarn lint
```

### Storybook

> We use [Storybooks](https://storybooks.js.org) for local development. You can use the command

Reviewer: ```suggestion
We use [Storybook](https://storybooks.js.org) for local development. Run the following command to start it:
```

```bash
yarn start
```

> followed by opening [http://localhost:9003](http://localhost:9003) in your browser to play around with the components and test your changes.

Reviewer: ```suggestion
Then, open [http://localhost:9003](http://localhost:9003) in your browser to play around with the components and test your changes.
```

### Documentation
Our documentation should always remain up to date. When making changes to components, make sure the appropriate documentation has been updated to reflect those changes. Documentation for each component can be found in the docs folder within a component's package. Other documentation pages can be found in the [packages/dev/docs](https://github.com/adobe-private/react-spectrum-v3/tree/master/packages/dev/docs) folder in the codebase.

Documentaton can be run locally by using
```bash
yarn start:docs
```

> followed by opening [http://localhost:1234](http://localhost:1234) in your browser.

Reviewer: ```suggestion
Then, open [http://localhost:1234](http://localhost:1234) in your browser.
```

## Contributor to committer

> We love contributions from our community! If you'd like to go a step beyond contributor and become a committer with full write access and a say in the project, you must be invited to the project. The existing committers employ an internal nomination process that must reach lazy consensus (silence is approval) before invitations are issued. If you feel you are qualified and want to get more deeply involved, feel free to reach out to existing committers to have a conversation.

Reviewer: ```suggestion
We love contributions from our community! If you'd like to go a step beyond contributor and become a committer with full write access and a say in the project, you must be invited. The existing committers employ an internal nomination process that must reach lazy consensus (silence is approval) before invitations are issued. If you feel you are qualified and want to get more deeply involved, feel free to reach out to existing committers to have a conversation.
```
